### DISCUSSION TITLE
How to patch Symbols Only Nerd Font

### DISCUSSION THREAD

**Discussion by lukejanicke** [ryanoasis/nerd-fonts#1774] [Q&A] [answered]
I would like to patch my own Symbols Nerd Font (the font file with just the Nerd Font symbols).

Is `src/glyphs/original-source.otf` the “font” to patch to create Symbols Nerd Font?

#479 discussed this question but closed after resolving a missing symbol.

`src/unpatched-fonts/NerdFontsSymbolsOnly` looks relevant but I don’t understand those files.

**Comment by Finii** [comment, marked as answer]
The Symbols Only font starts - like all other patched fonts - with a source font where all the symbols are patched into.
That source font is empty, so that in the end we have only all the icons in one font.
To give a somehow nice result the empty font needs to have some properties (like glyph size etc) defined, although it is empty.
To make maintenance more easy the font format is not `ttf` or `otf` but it is `sfd`

Empty source font `src/unpatched-fonts/NerdFontsSymbolsOnly/NerdFontsSymbolsNerdFontBlank.sfd`

The directory `src/glyphs/` contains all the individual icon sets. They are not scaled and possibly unsorted. The raw icons.

To patch the Symbols only font you need something like

```console
$ fontforge font-patcher --complete --ext ttf --variable-glyph-width src/unpatched-fonts/NerdFontsSymbolsOnly/NerdFontsSymbolsNerdFontBlank.sfd
```

You need to specify the `--ext` because probably you want a `ttf` (or `otf`) font as result and not a `sfd`. The patcher keeps the source font's type for the patched font if nothing is specified.

The `config.cfg` file you find with some unpatched source fonts (like the Symbols Only font) specifies which options need to be given for patching of that font; the automated release process uses this for example. So that is a good place to look for special options needed for that specific font.

**Reply by lukejanicke** [reply]
It worked! (Marked as answer.)

Well, I am doing the Docker thing so I grabbed `NerdFontsSymbolsNerdFontBlank.sfd` and ran:

```shell
./font-patcher -c --ext ttf fonts/NerdFontsSymbolsNerdFontBlank.sfd
```

**Reply by lukejanicke** [reply]
Footnote...

`./font-patcher -c --ext ttf fonts/NerdFontsSymbolsNerdFontBlank.sfd`
creates **SymbolsNerdFont-Regular.ttf**.

`./font-patcher -c -s --ext ttf fonts/NerdFontsSymbolsNerdFontBlank.sfd`
creates **SymbolsNerdFontMono-Regular.ttf** as expected.

But `./font-patcher -c --variable-width-glyphs --ext ttf fonts/NerdFontsSymbolsNerdFontBlank.sfd`
creates **SymbolsNerdFont-Regular.ttf** again and not a **Propo** variant as I was expecting.

My naive guess is that since we are not patching to a monospaced font, there is no need to distinguish `--variable-width-glyphs` from **Regular**. They should be the same. Is that right?

**Reply by lukejanicke** [reply]
Follow-up...

I used `fonttools` to change the name of the `--variable-width-glyphs` font to **Symbols Nerd Font Propo** so I could install it alongside **Symbols Nerd Font** and **Symbols Nerd Font Mono**. A visual side-by-side comparison shows that they are not the same.

This suggests the patcher is correctly patching `NerdFontsSymbolsNerdFontBlank.sfd` with **Propo** symbols but not naming the generated font file accordingly.

<img width="734" alt="Screenshot 2025-01-01 at 6 46 13 pm" src="https://github.com/user-attachments/assets/d9de9c74-bfc9-4207-a3d0-89f1323bdd36" />

**Reply by Finii** [reply]
The Symbols Only font comes in two flavours
* Nerd Font Mono
* Nerd Font Propo (which is called just `Nerd Font` only for this special case for backward compatibility and brevity)
* There is no real "Nerd Font" variant (i.e. were the icons extend into the next cell)

As I said, you need to give `--variable-width-glyphs`, for the normal Symbols Only font, or `--mono` (`-s`) for the Nerd Font Mono variant.

I'm not sure what giving neither produces for a font.

See also #1103 for an explanation of the three possible variants.

> They should be the same. Is that right?

This is not entirely clear to me, it might differ in subtile ways, or be identical. I guess they differ. It comes also into play that the Symbols Only font qualifies as 'short-and-stubby' font which is patched differently from fonts of usual aspect ratio cells (where the cell is higher than it is wide).

If you are curious, just create both (with no options and with `--variable`), open both in fontforge and run a font-diff.

**Reply by Finii** [reply]
> This suggests the patcher is correctly patching NerdFontsSymbolsNerdFontBlank.sfd with Propo symbols but not naming the generated font file accordingly

Yes, there is a specific rule for that in the patcher, lets see if I can find it...

**Reply by Finii** [reply]
Its here in this commit:
a7ecf075d2d9d50215d3a65129747a2bff3dda63

**Reply by Finii** [reply]
Additional info:

Releases (or people) use the `gotta-patch-em` script to run the patcher script, and that does in fact patch the Symbols Only font three times, but trashes the version without any options, so that in the end we have only the Mono and the not-called-Propo-but-it-is-Propo one.

To make THAT happen we have the special rule added with this commit: 63db683b1a6c4a7341ec5872f437a8953a321bb0

```python
if args.nonmono and args.single:
    logger.warning("Specified contradicting --variable-width-glyphs and --use-single-width-glyph. Ignoring --variable-width-glyphs.")
    args.nonmono = False
```

**Reply by ningw42** [reply]
I have a dumb question. Is it possible to "center" glyphs horizontally for SymbolsNerdFont-Regular.ttf based on another font's width? This may sound dumb for experts. But here is what I want. 

https://github.com/ryanoasis/nerd-fonts/issues/1330 this is exactly what I want. I can patch a font to put the nerd font glyphs at the horizontal center (of two cells? I guess). Now, I want to do it without a real font as the input <sup>1</sup>. I want the SymbolsNerdFont-Regular.ttf to carry that "centered" glyphs. I have zero knowledge of typography, but I can tell that the blank font `NerdFontsSymbolsNerdFontBlank.sfd` misses the width information of a real font we are going to work with, to calculate the center position. So, I am thinking of if I could take the necessary metadata (width, being the major or only one, I guess?) from a real font, fill in them in `NerdFontsSymbolsNerdFontBlank.sfd` (or maybe just pass to the patcher as an input), then apply the patch in https://github.com/ryanoasis/nerd-fonts/issues/1330. Will such a font-patcher generate a SymbolsNerdFont-Regular.ttf (when used as a fallback) shows glyph at horizontal center as if I patched the real font?

1. Why would I need that when I can just patch a real font?
    - I have a font that has a long long name which exceeds the 31 chars limit (`ERROR: ====-< Family (ID 1)      too long (39 > 31):`). when patched, their name gets truncated, making them hard to identify. btw, I see this is a validation in font-patcher, is this a real limitation somewhere else? the original font I have already exceeds that limit and it seems fine.
    - I can patch it with `--makegroups -1` to keep their original name. I do want to keep NF and the original font seperate. 
    - So, this leaves me no choice other than the question in the beginning. Generate a customized SymbolsNerdFont as a fallback font.


Thank you!

**Reply by Finii** [reply]
> "center" glyphs horizontally for SymbolsNerdFont-Regular.ttf based on another font's width

You should be able to specify a target 'cell' size when patching the SymbolsOnly font with `--cell 0:1000:-200:800`, where the numbers are just examples. The 'cell' size of your other font you can determine by "patching" that other font with options `--cell '?' --dry`. That will not patch just analyze and show you the size. Note the questionmark needs to be quoted on usual shells.

```console
$ fontforge font-patcher --cell '?' --dry Arimo-Regular.ttf
Nerd Fonts Patcher v3.4.0-86 (4.22.0) (ff 20230101)
INFO: Redistributing line gap of 67 (33 top and 34 bottom)
INFO: Cell coordinates (Xmin:Xmax:Ymin:Ymax) 0:2079:-468:1887
Done with Patch Sets, generating font...
$
```

I'm not sure the `--cell` actually works for the SymbolsOnly font as that font is handled differently from all others, in some aspects (and I cant remember which aspects).

Ah, that only works when the EM is the same for your other font and the SymbolsOnly font :thinking: The "unit" is EM, so if the EM differs you need to scale the `--cell` numbers accordingly. SymbolsOnly has 2048 units per EM.

> Will such a font-patcher generate a SymbolsNerdFont-Regular.ttf (when used as a fallback) shows glyph at horizontal center as if I patched the real font?

Well, how fallback-fonts are handled depends on multiple things. With normal font rendering you would just need to patch SymbolsOnly with your desired cell-width (which is 2 times the other font's width), and then the glyph is just extrawide.
Special terminals that use their own algorithms (and GPU) to render extra quickly have their own ideas how to render fonts and it will depend on them. I have no experience here.

> when patched, their name gets truncated

I do not remember any code that truncates the names :thinking: It will just dutifully create the font with the too long name. You can shorten the `Nerd Font` addition by specifying `--makegroups 4` for example.
I principle you do not need to bother with font name length except for Word-For-Windows 1.0 on Mac (:wink: or similar), and then recently I have read that Windows STILL looks only on that number of characters in the name to distinguish multiple installed fonts :unamused:.

_Edit:_

Options "explained" here: https://github.com/ryanoasis/nerd-fonts/wiki/ScriptOptions

**Reply by Finii** [reply]
Regarding the "centered within two cells" problem, I guess you can - instead of applying some patch - patch like this (untested):

```console
$ fontforge font-patcher --cell '?' --dry Arimo-Regular.ttf
Nerd Fonts Patcher v3.4.0-86 (4.22.0) (ff 20230101)
INFO: Redistributing line gap of 67 (33 top and 34 bottom)
INFO: Cell coordinates (Xmin:Xmax:Ymin:Ymax) 0:2079:-468:1887
Done with Patch Sets, generating font...
$ dc -e '2079 2 * p'
4158
$ fontforge font-patcher --cell 0:4158:-468:1887 --single-width-glyphs Arimo-Regular.ttf
...
```

_Edit:_
Ah, the resulting font is not monospaced anymore, but I guess maybe you want the Powerline glyphs to be single-width anyhow, so then thereafter run the font (again) through font-patcher with `--mono --powerline --powerlineextra` to a) make it monospaced and b) recreate the Powerline glyphs with one cell width.
Use `--makegroups -1` for the first run and `--makegroups 4` for the 2nd.

**Reply by Finii** [reply]
Hmm, another possibility instead of applying a patch could be (untested) to edit the `font-patcher` like this

```diff
         SYM_ATTR_DEFAULT = {
-            'default': {'align': 'c', 'valign': 'c', 'stretch': 'pa', 'params': {}}
+            'default': {'align': 'c', 'valign': 'c', 'stretch': 'pa2!', 'params': {}}

         }
```

That should in principle force (`!`) the glyphs to be two cells wide (`2`).
If you try this, please report if it works or not.

In this way you can select the width per-icon-set :sunglasses:

**Reply by ningw42** [reply]
Thank you for the detailed response. I didn't get a chance to try them out until now.

I tried the "pa2!" approach at the end, it didn't work exactly as the patch in https://github.com/ryanoasis/nerd-fonts/issues/1330.

This is how "pa2!" looks like in Windows Terminal.
<img width="1494" height="539" alt="image" src="https://github.com/user-attachments/assets/040706ea-e509-4ffb-919b-0777c110c2a8" />
<img width="422" height="53" alt="image" src="https://github.com/user-attachments/assets/82f52a05-2be8-456d-9f22-c4b2d1d5f102" />


This is how https://github.com/ryanoasis/nerd-fonts/issues/1330's patch looks like in Windows Terminal.
<img width="1505" height="497" alt="image" src="https://github.com/user-attachments/assets/09400809-72c9-4189-acb9-3a63bf35c2c8" />
<img width="425" height="46" alt="image" src="https://github.com/user-attachments/assets/a02e3bea-c3ac-4955-913d-c9ab5b12a09b" />

Look at the icons for `.pre-commit-config.yaml`, `.sops.yaml` and git branch. They are larger than other icons' "normal" size, and not centered horizontally.

**Reply by ningw42** [reply]
If I combine https://github.com/ryanoasis/nerd-fonts/issues/1330 's patch with "pa2!"
<img width="1500" height="498" alt="image" src="https://github.com/user-attachments/assets/0cabfe32-9fe4-453c-8089-deba0a53ffde" />

The icons for `.pre-commit-config.yaml` and `.sops.yaml` are now centered horizontally, but still larger than without "pa2!". Also, with "pa2!", other icons are larger. I guess that's the expected outcome?

**Reply by ningw42** [reply]
Hi, @Finii, 

It took me a while to understand this and have a working solution (at least to my eyes), with the help of today's AI. Thank you again for the detailed explanation. 

**TL;DR: creating a SymbolsNerdFont by specifying `--cell` with scaled values from the target source font.**

If I understanding correctly, https://github.com/ryanoasis/nerd-fonts/discussions/1774#discussioncomment-15694170 this should work without the patch in https://github.com/ryanoasis/nerd-fonts/issues/1330 when a patched font is the goal. And the `--cell` patch below is just to delay the sanity check that would block our patch on the empty font.






# AI Generated Content

## Building a Centered Two-Cell SymbolsNerdFont for BerkeleyMono-ExtraCondensed

### Background

[Nerd Fonts](https://github.com/ryanoasis/nerd-fonts) provides `SymbolsNerdFont` as a standalone fallback font containing only nerd-font icon glyphs. Terminal emulators can use it alongside any text font without patching the text font directly.

The nerd-fonts project's `font-patcher` supports a `--cell` flag (see [discussion #1774](https://github.com/ryanoasis/nerd-fonts/discussions/1774)) to override cell dimensions when patching a source font.

### Issue

Two bugs prevent this from working:

#### 1. `--cell` override comes too late (fatal exit on blank font)

In `font-patcher`, the `width <= 0` fatal check runs *before* the `--cell` X override is applied. A blank SFD has no glyphs, so the detected width is zero, and the script exits before `--cell` can set the correct width:

```python
# Original order (broken):
if self.font_dim['width'] <= 0:          # line 1355 — exits here
    logger.critical("Can not detect sane font width")
    sys.exit(1)
if isinstance(self.args.cellopt, list):   # line 1358 — never reached
    ...
    self.font_dim['width'] = self.args.cellopt[1]
```

#### 2. Center-alignment formula ignores two-cell target width

The center-alignment code centers glyphs within one cell width, but two-cell glyphs need centering across `cell_width * target_width` (i.e., 2 cells):

```python
# Original (centers within 1 cell):
x_align_distance += (cell_width / 2) - (sym_dim['width'] / 2)
```

### Goal

Produce a `SymbolsNerdFont-Regular.ttf` where all nerd-font glyphs are horizontally centered across two terminal cells of BerkeleyMono-ExtraCondensed.

### Solution

Two patch files fix the bugs, and a build script applies them to a temporary copy of `font-patcher` (leaving the original untouched).

#### Patch 1: `cellopt-reorder.patch`

Moves the `--cell` override block before the `width <= 0` check:

```diff
--- a/font-patcher
+++ b/font-patcher
@@ -1352,15 +1352,15 @@
         if self.font_dim['width'] < self.font_dim['xmax']:
             logger.debug("Font has negative right side bearing in extended glyphs")
             self.font_dim['xmax'] = self.font_dim['width'] # In fact 'xmax' is never used
-        if self.font_dim['width'] <= 0:
-            logger.critical("Can not detect sane font width")
-            sys.exit(1)
         if isinstance(self.args.cellopt, list):
             logger.debug("Overriding cell X{%d:%d} with X{%d:%d}",
                 self.font_dim['xmin'], self.font_dim['xmin'] + self.font_dim['width'],
                 self.args.cellopt[0], self.args.cellopt[1])
             self.font_dim['xmin'] = self.args.cellopt[0]
             self.font_dim['xmax'] = self.args.cellopt[1]
             self.font_dim['width'] = self.args.cellopt[1]
+        if self.font_dim['width'] <= 0:
+            logger.critical("Can not detect sane font width")
+            sys.exit(1)
         if self.args.cellopt:
             logger.info("Cell coordinates (Xmin:Xmax:Ymin:Ymax) %s%d:%d:%d:%d",
                 '' if not isinstance(self.args.cellopt, list) else 'overridden with ',
```

#### Patch 2: `horizontal-centered.patch`

Multiplies the centering width by `get_target_width(stretch)` (1 for single-cell, 2 for double-cell glyphs):

```diff
--- a/font-patcher
+++ b/font-patcher
@@ -1641,7 +1641,7 @@
                     cell_width = self.font_dim['width']
                 if sym_attr['align'] == 'c':
                     # Center align
-                    x_align_distance += (cell_width / 2) - (sym_dim['width'] / 2)
+                    x_align_distance += (cell_width * self.get_target_width(stretch) / 2) - (sym_dim['width'] / 2)
                 elif sym_attr['align'] == 'r':
                     # Right align
                     # (not really supported with pa scaling and 2x stretch in NFP)
```

#### Build script: `build.sh`

```bash
#!/usr/bin/env bash
set -euo pipefail

SCRIPT_DIR="$(cd "$(dirname "$0")" && pwd)"
SOURCE_FONT="$SCRIPT_DIR/source_font/BerkeleyMono-ExtraCondensed.otf"
BLANK_SFD="$SCRIPT_DIR/NerdFontsSymbolsNerdFontBlank.sfd"
PATCHER="$SCRIPT_DIR/font-patcher"
OUTPUT_DIR="$SCRIPT_DIR/output"

BLANK_EM=2048

# Probe the source font's cell width, EM, and vertical metrics
echo "Probing source font..."
read -r SRC_WIDTH SRC_EM SRC_TYPO_ASCENT SRC_TYPO_DESCENT < <(
    fontforge -quiet -lang=py -c "
import fontforge
f = fontforge.open('$SOURCE_FONT')
g = f[ord('M')]
print(g.width, f.em, f.os2_typoascent, f.os2_typodescent)
f.close()
" 2>/dev/null
)
echo "  Source font: cell width=$SRC_WIDTH, EM=$SRC_EM, typo ascent=$SRC_TYPO_ASCENT, typo descent=$SRC_TYPO_DESCENT"

# Scale source font metrics to the blank SFD's EM
SCALED_WIDTH=$(python3 -c "print(int(round($SRC_WIDTH * $BLANK_EM / $SRC_EM)))")
SCALED_ASCENT=$(python3 -c "print(int(round($SRC_TYPO_ASCENT * $BLANK_EM / $SRC_EM)))")
SCALED_DESCENT=$(python3 -c "print(int(round($SRC_TYPO_DESCENT * $BLANK_EM / $SRC_EM)))")
echo "  Scaled to EM=$BLANK_EM: width=$SCALED_WIDTH, ascent=$SCALED_ASCENT, descent=$SCALED_DESCENT"

# Create a patched copy of font-patcher
PATCHED_PATCHER=$(mktemp "$SCRIPT_DIR/.font-patcher-patched.XXXXXX")
trap 'rm -f "$PATCHED_PATCHER"' EXIT

cp "$PATCHER" "$PATCHED_PATCHER"
patch -s "$PATCHED_PATCHER" "$SCRIPT_DIR/cellopt-reorder.patch"
patch -s "$PATCHED_PATCHER" "$SCRIPT_DIR/horizontal-centered.patch"
echo "Patched font-patcher created."

# Run the patched font-patcher
mkdir -p "$OUTPUT_DIR"
echo "Running font-patcher..."
python3 "$PATCHED_PATCHER" \
    "$BLANK_SFD" \
    --cell "0:${SCALED_WIDTH}:${SCALED_DESCENT}:${SCALED_ASCENT}" \
    --complete \
    -ext ttf \
    --outputdir "$OUTPUT_DIR"

echo "Done. Output in $OUTPUT_DIR/"
ls -la "$OUTPUT_DIR/"
```

**Reply by Finii** [reply]
Although I hate this AI slop, I must admit that at least patch 1 is correct ;-)

The options have never been intended for the problems we have at hand now, and in general the swapped evaluation is more useful. I will change that shortly.

The second patch, I will also examine that and maybe apply.
I will list you as author with a link to this thread (and a reference to LLM).

Thanks for your determination :-)

**Reply by Finii** [reply]
The second patch also looks convincing, especially as I do the same already in the right-align version.
Have no clue why it is not done in the center-align code. Good catch LLM.

https://github.com/ryanoasis/nerd-fonts/blob/1514941a80397d93361f4193346d9cbb9ed21c6e/font-patcher#L1692-L1707

_Scroll down code block_

Line `1704` takes the target width into account, but line `1700` is missing the factor (as pointed out by you).

But I need to run some tests first to see if something has been overlooked 😬

**Reply by Finii** [reply]
`# (not really supported with pa scaling and 2x stretch in NFP)`

🤔 But NFP is not a goal here, I guess. But then, SymbolsOnly which looks like NF is in fact a NFP.
Ah but the aim is a NFM in this case 😌 

NF = Nerd Font (monospaced but glyphs hang into next cell)
NFM = Nerd Font Mono (strictly monospaced)
NFP = Nerd Font Propo (icons take the space they need, _not_ monospaced)

**Reply by Finii** [reply]
See
* #1992

**Reply by Finii** [reply]
Thank you for the contribution and also for the honesty regarding your LLM use. 👍

In recent [PR 1980](https://www.github.com/ryanoasis/nerd-fonts/pull/1980) I had the feeling I talked to an LLM (with the text formatting etc) but that has never been acknowledged. That felt somehow wrong. How you handled this I guess this feels ok.
